File: completed/0386.java
import java.util.*;

class Solution {
    private List<Integer> helper(int current, int max) {
        List<Integer> results = new ArrayList<>();
        if (current > max) {
            return results;
        }

        results.add(current);
        results.addAll(helper(current * 10, max));
        for (int i = 1; i <= 9; i++) {
            if (current + i > max) {
                break;
            }
            results.add(current + i);
            results.addAll(helper((current + i) * 10, max));
        }
        return results;
    }

    public List<Integer> lexicalOrder(int n) {
        List<Integer> results = new ArrayList<>();
        for (int i = 1; i <= 9; i++) {
            if (i > n) {
                break;
            }
            results.add(i);
            results.addAll(helper(i * 10, n));
        }
        return results;
    }
}

File: completed/0386.cpp
class Solution {
private:
    void add(int current, int n, vector<int> &result) {
        int start = current == 0 ? 1 : 0;
        for (int i = start; i <= 9; i++) {
            int next = current * 10 + i;
            if (next > n) {
                break;
            }
            result.push_back(next);
            add(next, n, result);
        }
    }

public:
    vector<int> lexicalOrder(int n) {
        vector<int> result;
        add(0, n, result);
        return result;
    }
};
